### ISSUE TITLE
Packed modules: miscommunication emacs/ocamldebug

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7540]
**Original bug ID:** 7540
**Reporter:** herbelin
**Status:** acknowledged (set by @xavierleroy on 2017-10-10T12:29:17Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Version:** 4.04.1
**Category:** tools (ocaml{lex,yacc,dep,debug,...})

## Bug description

[ Original summary was: Packed modules: setting "break" broken in ocamldebug + miscommunication emacs/ocamldebug.  First issue is solved. ]

There seems to be a bug with ocamldebug on packed modules. In the reproducible script below, the first "break" works but the second gives "Can't find any event there". Note however that executing step by step works well.

This seems related to the call to "Hashtbl.add all_events_by_module md sorted_evl" in Symbols.read_symbols which binds all modules of a pack to the name of the first module.

There is also a communication bug when calling ocamldebug from emacs (at least using tuareg). Commands of the form `break @ "b" # 10` are sent while ocamldebug tolerance is only about adding .ml or uncapitalizing the first letter of a (possibly qualified) module name.

## Steps to reproduce

echo "let f x = x" > a.ml
echo "open A;; f 1" > b.ml
ocamlc -g -c a.ml b.ml
ocamlc -g -pack -o p.cmo a.cmo b.cmo
ocamlc -g -o a.out p.cmo
ocamldebug a.out
break @ B.P # 10
break @ A.P # 9

## Additional information

Since I already tried a fix for testing, I shall try to submit it to the PR system in case it can ever save time.

**Comment by vicuna** [comment]
**Comment author:** @mshinwell

Please submit a Github PR with your fix (https://github.com/ocaml/ocaml), thanks.

**Comment by vicuna** [comment]
**Comment author:** @xavierleroy

Hugo, we're still interested in your code (the one that attempts to fix the loading of events in ocamldebug).  Having your code as a possible fix or at least a starting point would help us a lot.  Thank you.

**Comment by vicuna** [comment]
**Comment author:** herbelin

Hi Xavier, regarding the fix to ocamldebug, I had submitted https://github.com/ocaml/ocaml/pull/1179 which has been merged on Jul 19 (https://github.com/ocaml/ocaml/commit/23ba4788f0a30e4c64565e9a08bc6557b58b95bf).

I had another issue with tuareg which does not support sending breakpoints to ocamldebug when modules are in a pack. I reported this at https://github.com/ocaml/tuareg/issues/136 but I have no clue how to fix that, since either tuareg should get a way to know that some files are in a pack, or ocamldebug being able to recover the fully qualified module name from the base name, but this would somehow defeat the purpose of packing.

Actually, the issue seems to exist also for the version of camldebug.el shipped with ocaml (https://github.com/ocaml/ocaml/blob/019f469aa36c65d5436c0cc107aff310aa399735/emacs/camldebug.el#L190).
